Build failure: ucc with CUDA on unstable-small

## Steps To Reproduce

Steps to reproduce the behavior:

1. build *ucc*

## Build log

<!-- insert build log in code block in collapsable section -->

<details>

<summary>Build Log</summary>

```
warning: The interpretation of store paths arguments ending in `.drv` recently changed. If this command is now failing try again with '/nix/store/1dn6nk9q1rd512ip5dxc8m0c4l8igxxg-ucc-1.3.0.drv^*'
got build log for '/nix/store/1dn6nk9q1rd512ip5dxc8m0c4l8igxxg-ucc-1.3.0.drv' from 'daemon'
Sourcing setup-cuda-hook
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/j2hqdb0dxjkrmavba0q1y83lb918234l-source
source root is source
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
substituteStream() in derivation ucc-1.3.0: WARNING: '--replace' is deprecated, use --replace-{fail,warn,quiet}. (file 'src/components/tl/mlx5/Makefile.am')
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/tl/mlx5/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/tl/sharp/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/tl/ucp/coll_plugins/example/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/tl/ucp/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/tl/nccl/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/tl/rccl/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/tl/cuda/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/tl/self/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/cl/hier/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/cl/basic/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/ec/cpu/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/ec/cuda/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/ec/rocm/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/mc/cpu/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/mc/cuda/Makefile.am'
substituteStream() in derivation ucc-1.3.0: WARNING: pattern /bin/bash doesn't match anything in file 'src/components/mc/rocm/Makefile.am'
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
autoreconf: export WARNINGS=
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force 
sh: line 1: git: command not found
sh: line 1: git: command not found
aclocal: warning: couldn't open directory 'm4': No such file or directory
autoreconf: configure.ac: tracing
sh: line 1: git: command not found
sh: line 1: git: command not found
autoreconf: running: libtoolize --copy --force
libtoolize: putting auxiliary files in '.'.
libtoolize: copying file './ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
libtoolize: Consider adding '-I m4' to ACLOCAL_AMFLAGS in Makefile.am.
autoreconf: configure.ac: not using Intltool
autoreconf: configure.ac: not using Gtkdoc
autoreconf: running: aclocal --force 
sh: line 1: git: command not found
sh: line 1: git: command not found
autoreconf: running: /nix/store/m0knd99ncm7kbpffm10gsdmgv1j6mgy6-autoconf-2.72/bin/autoconf --force
sh: line 1: git: command not found
sh: line 1: git: command not found
autoreconf: running: /nix/store/m0knd99ncm7kbpffm10gsdmgv1j6mgy6-autoconf-2.72/bin/autoheader --force
sh: line 1: git: command not found
sh: line 1: git: command not found
autoreconf: running: automake --add-missing --copy --force-missing
configure.ac:31: installing './compile'
configure.ac:82: installing './config.guess'
configure.ac:82: installing './config.sub'
configure.ac:50: installing './install-sh'
configure.ac:50: installing './missing'
src/Makefile.am: installing './depcomp'
autoreconf: './install-sh' is updated
autoreconf: Leaving directory '.'
Executing setupCUDAToolkitCompilers
fixing libtool script ./ltmain.sh
./configure
patching script interpreter paths in ./configure
./configure: interpreter directive changed from "#! /bin/sh" to "/nix/store/5mh7kaj2fyv8mk4sfq1brwxgc02884wi-bash-5.2p37/bin/sh"
configure flags: --disable-static --disable-dependency-tracking --prefix=/nix/store/p01vgbfj4xqrq6nrqk0jbmfg4sc7zvi5-ucc-1.3.0 --bindir=/nix/store/p01vgbfj4xqrq6nrqk0jbmfg4sc7zvi5-ucc-1.3.0/bin --sbindir=/nix/store/p01vgbfj4xqrq6nrqk0jbmfg4sc7zvi5-ucc-1.3.0/sbin --includedir=/nix/store/l4dzz9313syax1kis0rjy1k1775cxipa-ucc-1.3.0-dev/include --oldincludedir=/nix/store/l4dzz9313syax1kis0rjy1k1775cxipa-ucc-1.3.0-dev/include --mandir=/nix/store/p01vgbfj4xqrq6nrqk0jbmfg4sc7zvi5-ucc-1.3.0/share/man --infodir=/nix/store/p01vgbfj4xqrq6nrqk0jbmfg4sc7zvi5-ucc-1.3.0/share/info --docdir=/nix/store/p01vgbfj4xqrq6nrqk0jbmfg4sc7zvi5-ucc-1.3.0/share/doc/ucc --libdir=/nix/store/p01vgbfj4xqrq6nrqk0jbmfg4sc7zvi5-ucc-1.3.0/lib --libexecdir=/nix/store/p01vgbfj4xqrq6nrqk0jbmfg4sc7zvi5-ucc-1.3.0/libexec --localedir=/nix/store/p01vgbfj4xqrq6nrqk0jbmfg4sc7zvi5-ucc-1.3.0/share/locale --with-cuda=/nix/store/2l3lv7yha1gqvdyig02cpr0mh13cc7sb-cuda_cudart-12.4.99 --with-nvcc-gencode=-gencode=arch=compute_60\,code=sm_60\ -gencode=arch=compute_61\,code=sm_61\ -gencode=arch=compute_70\,code=sm_70\ -gencode=arch=compute_75\,code=sm_75\ -gencode=arch=compute_80\,code=sm_80\ -gencode=arch=compute_86\,code=sm_86\ -gencode=arch=compute_89\,code=sm_89\ -gencode=arch=compute_90\,code=sm_90\ -gencode=arch=compute_90a\,code=sm_90a\ -gencode=arch=compute_90a\,code=compute_90a
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking whether gcc understands -c and -o together... yes
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for wchar.h... yes
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking whether _XOPEN_SOURCE should be defined... no
configure: builddir: /build/source
configure: srcdir: /build/source
checking for a BSD-compatible install... /nix/store/4s9rah4cwaxflicsk5cndnknqlk9n4p3-coreutils-9.5/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /nix/store/4s9rah4cwaxflicsk5cndnknqlk9n4p3-coreutils-9.5/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports the include directive... yes (GNU style)
checking whether make supports nested variables... yes
checking whether UID '1000' is supported by ustar format... yes
checking whether GID '100' is supported by ustar format... yes
checking how to create a ustar tar archive... gnutar
checking dependency style of gcc... none
checking whether make supports nested variables... (cached) yes
checking whether to enable maintainer-specific portions of Makefiles... no
configure: git sha: 
checking for gcc... (cached) gcc
checking whether the compiler supports GNU C... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to enable C11 features... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking whether the compiler supports GNU C++... yes
checking whether g++ accepts -g... yes
checking for g++ option to enable C++11 features... none needed
checking dependency style of g++... none
checking dependency style of gcc... none
checking whether ln -s works... yes
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking how to print strings... printf
checking for a sed that does not truncate output... /nix/store/abm77lnrkrkb58z6xp1qwjcr1xgkcfwm-gnused-4.9/bin/sed
checking for grep that handles long lines and -e... /nix/store/aap6cq56amx4mzbyxp2wpgsf1kqjcr1f-gnugrep-3.11/bin/grep
checking for egrep... /nix/store/aap6cq56amx4mzbyxp2wpgsf1kqjcr1f-gnugrep-3.11/bin/grep -E
checking for fgrep... /nix/store/aap6cq56amx4mzbyxp2wpgsf1kqjcr1f-gnugrep-3.11/bin/grep -F
checking for ld used by gcc... ld
checking if the linker (ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... nm
checking the name lister (nm) interface... BSD nm
checking the maximum length of command line arguments... 1572864
checking how to convert x86_64-pc-linux-gnu file names to x86_64-pc-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for ld option to reload object files... -r
checking for file... file
checking for objdump... objdump
checking how to recognize dependent libraries... (cached) pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse nm output from gcc object... ok
checking for sysroot... no
checking for a working dd... /nix/store/4s9rah4cwaxflicsk5cndnknqlk9n4p3-coreutils-9.5/bin/dd
checking how to truncate binary pipes... /nix/store/4s9rah4cwaxflicsk5cndnknqlk9n4p3-coreutils-9.5/bin/dd bs=4096 count=1
checking for mt... no
checking if : is a manifest tool... no
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no
checking how to run the C++ preprocessor... g++ -E
checking for ld used by g++... ld -m elf_x86_64
checking if the linker (ld -m elf_x86_64) is GNU ld... yes
checking whether the g++ linker (ld -m elf_x86_64) supports shared libraries... yes
checking for g++ option to produce PIC... -fPIC -DPIC
checking if g++ PIC flag -fPIC -DPIC works... yes
checking if g++ static flag -static works... no
checking if g++ supports -c -o file.o... yes
checking if g++ supports -c -o file.o... (cached) yes
checking whether the g++ linker (ld -m elf_x86_64) supports shared libraries... yes
checking dynamic linker characteristics... (cached) GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking size of float... 4
checking size of double... 8
checking size of long double... 16
checking size of float _Complex... 8
checking size of double _Complex... 16
checking size of long double _Complex... 32
checking if ln -s supports --relative... yes
checking for doxygen... no
configure: WARNING: doxygen not found - will not generate any doxygen documentation
./configure: line 19832: doxygen: command not found
configure: WARNING: doxygen version  is bad. Required version: 1.8.6 and above
checking for perl... no
checking for g++ options needed to detect all undeclared functions... none needed
checking for size_t... yes
checking for gcc options needed to detect all undeclared functions... none needed
checking compiler flag -diag-error 10006... no
checking compiler flag -diag-error 10148... no
checking whether -diag-disable 1478 overrides deprecated declarations... no
checking whether -Wno-deprecated-declarations overrides deprecated declarations... yes
checking compiler flag -diag-disable 269... no
configure: Detected CPU implementation: 
configure: Detected CPU architecture: 
configure: Detected CPU variant: 
configure: Detected CPU part: 
checking for __attribute__(optimize)... 1
checking compiler flag -funwind-tables... yes
configure: compiling with unwind tables
checking if g++ works... yes
checking c++11 support... yes
checking gnu++11 support... yes
checking whether _GLIBCXX_NOTHROW is declared... yes
checking compiler flag --display_error_number... no
checking compiler flag --diag_suppress 181... no
checking compiler flag --diag_suppress 381... no
checking compiler flag --diag_suppress 1215... no
checking compiler flag --diag_suppress 1901... no
checking compiler flag --diag_suppress 1902... no
checking compiler flag --diag_suppress 301... no
checking compiler flag -pedantic... yes
checking compiler flag -Wno-missing-field-initializers... yes
checking compiler flag -Wno-unused-parameter... yes
checking compiler flag -Wno-long-long... yes
checking compiler flag -Wno-endif-labels... yes
checking compiler flag -Wno-sign-compare... yes
checking compiler flag -Wno-multichar... yes
checking compiler flag -Wno-deprecated-declarations... yes
checking compiler flag -Winvalid-pch... yes
checking compiler flag -Wno-pointer-sign... yes
checking compiler flag -Werror-implicit-function-declaration... yes
checking compiler flag -Wnested-externs... yes
checking compiler flag -Wshadow... yes
checking compiler flag -Wenum-conversion... yes
checking whether byte ordering is bigendian... no
checking for working alloca.h... yes
checking for alloca... yes
checking for shm_open in -lrt... yes
checking for timer_create in -lrt... yes
checking for libgen.h... yes
checking whether asprintf is declared... yes
checking whether basename is declared... yes
checking whether fmemopen is declared... yes
MPI perftest: disabled
checking for ucp/api/ucp.h... yes
checking for ucp_tag_send_nb in -lucp... yes
checking for ucs/sys/uid.h... yes
checking for ucs_get_system_id in -lucs... yes
Detected UCX version: 1.17
checking for ucs_mpool_params_t.ops... yes
checking for ucs_config_global_list_entry_t.flags... yes
checking for ucs_rcache_region_t.alignment... yes
checking for ucs_config_parser_t.doc... yes
UCX support: yes
checking for cuda.h... yes
checking for cuda_runtime.h... yes
checking for cuDeviceGetUuid in -lcuda... yes
checking for cudaGetDeviceCount in -lcudart... yes
checking for nvml.h... no
configure: WARNING: nvml header not found. Install appropriate cuda-nvml-devel package
checking size of cuFloatComplex... 8
checking size of cuDoubleComplex... 16
checking for nvcc... /nix/store/vnwm837kbwv0ckf1xbjs6zx9vddcjz8d-cuda_nvcc-12.4.99/bin/nvcc
Detected CUDA version: 12.4
CUDA support: yes; -I/nix/store/2l3lv7yha1gqvdyig02cpr0mh13cc7sb-cuda_cudart-12.4.99/include -L/nix/store/2l3lv7yha1gqvdyig02cpr0mh13cc7sb-cuda_cudart-12.4.99/lib -L/nix/store/2l3lv7yha1gqvdyig02cpr0mh13cc7sb-cuda_cudart-12.4.99/lib/stubs
configure: ROCm path was not specified. Guessing ...
checking for hsa.h... no
configure: WARNING: ROCm not found
checking if ROCm version is 6.0 or above... no
checking if ROCm version is 5.0 - 5.7... no
checking for hipFree in -lhip_hcc... no
checking for hip/hip_runtime.h... no
configure: using hipcc as ROCm version is 3.7.0 to ROCm 5.7.1
configure: WARNING: HIP Runtime not found
ROCM support: no; -I/opt/rocm/include/hsa -I/opt/rocm/include -L/opt/rocm/lib -L/opt/rocm/lib64 -L/opt/rocm/hsa/lib
HIP support: no; -D__HIP_PLATFORM_AMD__ -I/opt/rocm/include/hip -I/opt/rocm/include -I/opt/rocm/llvm/include -I/opt/rocm/hip/include -L/opt/rocm/lib -L/opt/rocm/llvm/lib -L/opt/rocm/hip/lib
GTEST support: 
checking for infiniband/verbs.h... no
configure: WARNING: ibibverbs header files not found
configure: WARNING: IBVERBS not found
IBVERBS support: no, MLX5DV support: 
checking for /usr/include/rdma/rdma_cma.h... no
configure: WARNING: rdmacm header files not found
configure: WARNING: rdmacm not found
RDMACM support: no
checking for nccl.h... no
configure: WARNING: NCCL not found
NCCL support: no
configure: WARNING: RCCL not found
RCCL support: no
checking for sharp/api/sharp_coll.h... no
configure: WARNING: SHARP not found
SHARP support: no
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating src/components/tl/cuda/Makefile
config.status: creating src/components/tl/mlx5/Makefile
config.status: creating src/components/tl/nccl/Makefile
config.status: creating src/components/tl/rccl/Makefile
config.status: creating src/components/tl/self/Makefile
config.status: creating src/components/tl/sharp/Makefile
config.status: creating src/components/tl/ucp/Makefile
config.status: creating src/components/tl/ucp/coll_plugins/example/Makefile
config.status: creating Makefile
config.status: creating src/Makefile
config.status: creating src/ucc/api/ucc_version.h
config.status: creating src/core/ucc_version.c
config.status: creating src/components/cl/basic/Makefile
config.status: creating src/components/cl/hier/Makefile
config.status: creating src/components/mc/cpu/Makefile
config.status: creating src/components/mc/cuda/Makefile
config.status: creating src/components/ec/cpu/Makefile
config.status: creating src/components/ec/cuda/Makefile
config.status: creating src/components/ec/cuda/kernel/Makefile
config.status: creating src/components/mc/rocm/Makefile
config.status: creating src/components/ec/rocm/Makefile
config.status: creating src/components/ec/rocm/kernel/Makefile
config.status: creating test/mpi/Makefile
config.status: creating test/gtest/Makefile
config.status: creating tools/info/Makefile
config.status: creating tools/perf/Makefile
config.status: creating cmake/Makefile
config.status: creating cmake/ucc-config-version.cmake
config.status: creating cmake/ucc-config.cmake
config.status: creating cmake/ucc-targets.cmake
config.status: creating config.h
config.status: executing depfiles commands
config.status: executing libtool commands
configure: =========================================================
configure: UCC build configuration:
configure:       Build prefix:   /nix/store/p01vgbfj4xqrq6nrqk0jbmfg4sc7zvi5-ucc-1.3.0
configure: Preprocessor flags:      -DCPU_FLAGS="" -I${abs_top_srcdir}/src -I${abs_top_builddir} -I${abs_top_builddir}/src
configure:         C compiler:   gcc  -O3 -g -DNDEBUG -std=gnu11 -Wall -Werror -funwind-tables -Wno-missing-field-initializers -Wno-unused-parameter -Wno-long-long -Wno-endif-labels -Wno-sign-compare -Wno-multichar -Wno-deprecated-declarations -Winvalid-pch -Wno-pointer-sign -Werror-implicit-function-declaration -Wnested-externs -Wshadow -Wenum-conversion
configure:       C++ compiler:   g++  -O3 -g -DNDEBUG -Wall -Werror -funwind-tables -Wno-missing-field-initializers -Wno-unused-parameter -Wno-long-long -Wno-endif-labels -Wno-sign-compare -Wno-multichar -Wno-deprecated-declarations -Winvalid-pch
configure:      NVCC gencodes:   -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_75,code=sm_75 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_89,code=sm_89 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90a,code=sm_90a -gencode=arch=compute_90a,code=compute_90a
configure:           Perftest:   disabled
configure:              Gtest:   disabled
configure:         MC modules:   < cpu cuda >
configure:         TL modules:   < self ucp >
configure:       TLCP modules:   < >
configure: =========================================================
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
build flags: -j12 SHELL=/nix/store/5mh7kaj2fyv8mk4sfq1brwxgc02884wi-bash-5.2p37/bin/bash
make  all-recursive
make[1]: Entering directory '/build/source'
Making all in src
make[2]: Entering directory '/build/source/src'
Making all in .
make[3]: Entering directory '/build/source/src'
  CC       core/libucc_la-ucc_lib.lo
  CC       core/libucc_la-ucc_constructor.lo
  CC       core/libucc_la-ucc_global_opts.lo
  CC       core/libucc_la-ucc_version.lo
  CC       core/libucc_la-ucc_context.lo
  CC       core/libucc_la-ucc_team.lo
  CC       core/libucc_la-ucc_ee.lo
  CC       core/libucc_la-ucc_coll.lo
  CC       core/libucc_la-ucc_progress_queue.lo
  CC       core/libucc_la-ucc_progress_queue_st.lo
  CC       core/libucc_la-ucc_progress_queue_mt.lo
  CC       core/libucc_la-ucc_service_coll.lo
  CC       core/libucc_la-ucc_dt.lo
  CC       schedule/libucc_la-ucc_schedule.lo
  CC       schedule/libucc_la-ucc_schedule_pipelined.lo
  CC       coll_score/libucc_la-ucc_coll_score.lo
  CC       coll_score/libucc_la-ucc_coll_score_map.lo
  CC       utils/libucc_la-ini.lo
  CC       utils/libucc_la-ucc_component.lo
  CC       utils/libucc_la-ucc_datastruct.lo
  CC       utils/libucc_la-ucc_status.lo
  CC       utils/libucc_la-ucc_mpool.lo
  CC       utils/libucc_la-ucc_math.lo
  CC       utils/libucc_la-ucc_proc_info.lo
  CC       utils/libucc_la-ucc_string.lo
  CC       utils/libucc_la-ucc_coll_utils.lo
  CC       utils/libucc_la-ucc_parser.lo
  CC       utils/profile/libucc_la-ucc_profile.lo
  CC       utils/libucc_la-ucc_sys.lo
  CC       utils/arch/x86_64/libucc_la-cpu.lo
  CC       utils/arch/aarch64/libucc_la-cpu.lo
  CC       utils/libucc_la-ucc_assert.lo
  CC       components/base/libucc_la-ucc_base_iface.lo
  CC       components/cl/libucc_la-ucc_cl.lo
  CC       components/tl/libucc_la-ucc_tl.lo
  CC       components/mc/libucc_la-ucc_mc.lo
  CC       components/mc/base/libucc_la-ucc_mc_base.lo
  CC       components/ec/libucc_la-ucc_ec.lo
  CC       components/ec/base/libucc_la-ucc_ec_base.lo
  CC       components/topo/libucc_la-ucc_topo.lo
  CC       components/topo/libucc_la-ucc_sbgp.lo
  CCLD     libucc.la
make[3]: Leaving directory '/build/source/src'
Making all in components/cl/basic
make[3]: Entering directory '/build/source/src/components/cl/basic'
  CC       libucc_cl_basic_la-cl_basic.lo
  CC       libucc_cl_basic_la-cl_basic_lib.lo
  CC       libucc_cl_basic_la-cl_basic_context.lo
  CC       libucc_cl_basic_la-cl_basic_team.lo
  CC       libucc_cl_basic_la-cl_basic_coll.lo
  CCLD     libucc_cl_basic.la
  LN       libucc_cl_basic.la
  LN       .libs/libucc_cl_basic.so
  LN       .libs/libucc_cl_basic.so.1
  LN       .libs/libucc_cl_basic.so.1.0.0
make[3]: Leaving directory '/build/source/src/components/cl/basic'
Making all in components/cl/hier
make[3]: Entering directory '/build/source/src/components/cl/hier'
  CC       libucc_cl_hier_la-cl_hier.lo
  CC       libucc_cl_hier_la-cl_hier_lib.lo
  CC       libucc_cl_hier_la-cl_hier_context.lo
  CC       libucc_cl_hier_la-cl_hier_team.lo
  CC       libucc_cl_hier_la-cl_hier_coll.lo
  CC       allreduce/libucc_cl_hier_la-allreduce.lo
  CC       allreduce/libucc_cl_hier_la-allreduce_rab.lo
  CC       allreduce/libucc_cl_hier_la-allreduce_split_rail.lo
  CC       alltoallv/libucc_cl_hier_la-alltoallv.lo
  CC       alltoall/libucc_cl_hier_la-alltoall.lo
  CC       barrier/libucc_cl_hier_la-barrier.lo
  CC       bcast/libucc_cl_hier_la-bcast.lo
  CC       bcast/libucc_cl_hier_la-bcast_2step.lo
  CCLD     libucc_cl_hier.la
  LN       libucc_cl_hier.la
  LN       .libs/libucc_cl_hier.so
  LN       .libs/libucc_cl_hier.so.1
  LN       .libs/libucc_cl_hier.so.1.0.0
make[3]: Leaving directory '/build/source/src/components/cl/hier'
Making all in components/mc/cpu
make[3]: Entering directory '/build/source/src/components/mc/cpu'
  CC       libucc_mc_cpu_la-mc_cpu.lo
  CCLD     libucc_mc_cpu.la
  LN       libucc_mc_cpu.la
  LN       .libs/libucc_mc_cpu.so
  LN       .libs/libucc_mc_cpu.so.1
  LN       .libs/libucc_mc_cpu.so.1.0.0
make[3]: Leaving directory '/build/source/src/components/mc/cpu'
Making all in components/mc/cuda
make[3]: Entering directory '/build/source/src/components/mc/cuda'
  CC       libucc_mc_cuda_la-mc_cuda.lo
  CCLD     libucc_mc_cuda.la
  LN       libucc_mc_cuda.la
  LN       .libs/libucc_mc_cuda.so
  LN       .libs/libucc_mc_cuda.so.1
  LN       .libs/libucc_mc_cuda.so.1.0.0
make[3]: Leaving directory '/build/source/src/components/mc/cuda'
Making all in components/ec/cpu
make[3]: Entering directory '/build/source/src/components/ec/cpu'
  CC       libucc_ec_cpu_la-ec_cpu.lo
  CC       libucc_ec_cpu_la-ec_cpu_reduce.lo
  CCLD     libucc_ec_cpu.la
  LN       libucc_ec_cpu.la
  LN       .libs/libucc_ec_cpu.so
  LN       .libs/libucc_ec_cpu.so.1
  LN       .libs/libucc_ec_cpu.so.1.0.0
make[3]: Leaving directory '/build/source/src/components/ec/cpu'
Making all in components/ec/cuda
make[3]: Entering directory '/build/source/src/components/ec/cuda'
Making all in kernel
make[4]: Entering directory '/build/source/src/components/ec/cuda/kernel'
/nix/store/5mh7kaj2fyv8mk4sfq1brwxgc02884wi-bash-5.2p37/bin/bash ../../../../../cuda_lt.sh "/nix/store/5mh7kaj2fyv8mk4sfq1brwxgc02884wi-bash-5.2p37/bin/bash ../../../../../libtool" ec_cuda_wait_kernel.lo /nix/store/vnwm837kbwv0ckf1xbjs6zx9vddcjz8d-cuda_nvcc-12.4.99/bin/nvcc -c  ec_cuda_wait_kernel.cu   -O3 -g -DNDEBUG -I/build/source -I/build/source/src -I/build/source/src --compiler-options -fno-rtti,-fno-exceptions -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_75,code=sm_75 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_89,code=sm_89 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90a,code=sm_90a -gencode=arch=compute_90a,code=compute_90a
/nix/store/5mh7kaj2fyv8mk4sfq1brwxgc02884wi-bash-5.2p37/bin/bash ../../../../../cuda_lt.sh "/nix/store/5mh7kaj2fyv8mk4sfq1brwxgc02884wi-bash-5.2p37/bin/bash ../../../../../libtool" ec_cuda_executor.lo /nix/store/vnwm837kbwv0ckf1xbjs6zx9vddcjz8d-cuda_nvcc-12.4.99/bin/nvcc -c  ec_cuda_executor.cu   -O3 -g -DNDEBUG -I/build/source -I/build/source/src -I/build/source/src --compiler-options -fno-rtti,-fno-exceptions -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_75,code=sm_75 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_89,code=sm_89 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90a,code=sm_90a -gencode=arch=compute_90a,code=compute_90a
/nix/store/5mh7kaj2fyv8mk4sfq1brwxgc02884wi-bash-5.2p37/bin/bash ../../../../../cuda_lt.sh "/nix/store/5mh7kaj2fyv8mk4sfq1brwxgc02884wi-bash-5.2p37/bin/bash ../../../../../libtool" ec_cuda_reduce.lo /nix/store/vnwm837kbwv0ckf1xbjs6zx9vddcjz8d-cuda_nvcc-12.4.99/bin/nvcc -c  ec_cuda_reduce.cu   -O3 -g -DNDEBUG -I/build/source -I/build/source/src -I/build/source/src --compiler-options -fno-rtti,-fno-exceptions -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_75,code=sm_75 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_89,code=sm_89 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90a,code=sm_90a -gencode=arch=compute_90a,code=compute_90a
/nix/store/vnwm837kbwv0ckf1xbjs6zx9vddcjz8d-cuda_nvcc-12.4.99/bin/nvcc -c ec_cuda_executor.cu -O3 -g -DNDEBUG -I/build/source -I/build/source/src -I/build/source/src --compiler-options -fno-rtti,-fno-exceptions -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_75,code=sm_75 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_89,code=sm_89 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90a,code=sm_90a -gencode=arch=compute_90a,code=compute_90a -Xcompiler -fPIC -o ./.libs/ec_cuda_executor.o
/nix/store/vnwm837kbwv0ckf1xbjs6zx9vddcjz8d-cuda_nvcc-12.4.99/bin/nvcc -c ec_cuda_wait_kernel.cu -O3 -g -DNDEBUG -I/build/source -I/build/source/src -I/build/source/src --compiler-options -fno-rtti,-fno-exceptions -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_75,code=sm_75 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_89,code=sm_89 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90a,code=sm_90a -gencode=arch=compute_90a,code=compute_90a -Xcompiler -fPIC -o ./.libs/ec_cuda_wait_kernel.o
/nix/store/vnwm837kbwv0ckf1xbjs6zx9vddcjz8d-cuda_nvcc-12.4.99/bin/nvcc -c ec_cuda_reduce.cu -O3 -g -DNDEBUG -I/build/source -I/build/source/src -I/build/source/src --compiler-options -fno-rtti,-fno-exceptions -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_75,code=sm_75 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_89,code=sm_89 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90a,code=sm_90a -gencode=arch=compute_90a,code=compute_90a -Xcompiler -fPIC -o ./.libs/ec_cuda_reduce.o
In file included from [01m[K/nix/store/pjwsmd212zhpriymn116xfl95zqxcwz5-cuda_cudart-12.4.99-dev/include/cuda_runtime.h:82[m[K,
                 from [01m[K<command-line>[m[K:
[01m[K/nix/store/vnwm837kbwv0ckf1xbjs6zx9vddcjz8d-cuda_nvcc-12.4.99/include/crt/host_config.h:143:2:[m[K [01;31m[Kerror: [m[K#error -- unsupported GNU version! gcc versions later than 13 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
  143 | #[01;31m[Kerror[m[K -- unsupported GNU version! gcc versions later than 13 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
      |  [01;31m[K^~~~~[m[K
In file included from [01m[K/nix/store/pjwsmd212zhpriymn116xfl95zqxcwz5-cuda_cudart-12.4.99-dev/include/cuda_runtime.h:82[m[K,
                 from [01m[K<command-line>[m[K:
[01m[K/nix/store/vnwm837kbwv0ckf1xbjs6zx9vddcjz8d-cuda_nvcc-12.4.99/include/crt/host_config.h:143:2:[m[K [01;31m[Kerror: [m[K#error -- unsupported GNU version! gcc versions later than 13 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
  143 | #[01;31m[Kerror[m[K -- unsupported GNU version! gcc versions later than 13 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
      |  [01;31m[K^~~~~[m[K
In file included from [01m[K/nix/store/pjwsmd212zhpriymn116xfl95zqxcwz5-cuda_cudart-12.4.99-dev/include/cuda_runtime.h:82[m[K,
                 from [01m[K<command-line>[m[K:
[01m[K/nix/store/vnwm837kbwv0ckf1xbjs6zx9vddcjz8d-cuda_nvcc-12.4.99/include/crt/host_config.h:143:2:[m[K [01;31m[Kerror: [m[K#error -- unsupported GNU version! gcc versions later than 13 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
  143 | #[01;31m[Kerror[m[K -- unsupported GNU version! gcc versions later than 13 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
      |  [01;31m[K^~~~~[m[K
make[4]: *** [Makefile:685: ec_cuda_wait_kernel.lo] Error 1
make[4]: *** Waiting for unfinished jobs....
make[4]: *** [Makefile:685: ec_cuda_reduce.lo] Error 1
make[4]: *** [Makefile:685: ec_cuda_executor.lo] Error 1
make[4]: Leaving directory '/build/source/src/components/ec/cuda/kernel'
make[3]: *** [Makefile:705: all-recursive] Error 1
make[3]: Leaving directory '/build/source/src/components/ec/cuda'
make[2]: *** [Makefile:1406: all-recursive] Error 1
make[2]: Leaving directory '/build/source/src'
make[1]: *** [Makefile:573: all-recursive] Error 1
make[1]: Leaving directory '/build/source'
make: *** [Makefile:499: all] Error 2
```

</details>

## Additional context

<!-- Add any other context about the problem here. -->

## Metadata

<!-- Please insert the output of running `nix-shell -p nix-info --run "nix-info -m"` below this line -->
```❯ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.4, NixOS, 25.05 (Warbler), 25.05.20241213.3566ab7`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.25.2`
 - nixpkgs: `/nix/store/22r7q7s9552gn1vpjigkbhfgcvhsrz68-source`
```

Trying to update my nixpkgs
```
• Updated input 'nixpkgs':
    'github:nixos/nixpkgs/3566ab7246670a43abd2ffa913cc62dad9cdf7d5?narHash=sha256-AKU6qqskl0yf2%2BJdRdD0cfxX4b9x3KKV5RqA6wijmPM%3D' (2024-12-13)
  → 'github:nixos/nixpkgs/adaa9f280329b5f814e8dc83eceddd42b20f72f4?narHash=sha256-EYaJtjLImMXzYEf9h52hkuPolXqKG/cAJlCOba6emL0%3D' (2024-12-29)
```

## Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->
@gebner
@srounce

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/uc/ucc/package.nix
+++ b/pkgs/by-name/uc/ucc/package.nix
@@ -1,4 +1,4 @@
-{
+inputs@{
   autoconf,
   automake,
   config,
@@ -15,6 +15,7 @@
   enableSse42 ? stdenv.hostPlatform.sse4_2Support,
 }:
 let
+  inherit (lib.attrsets) getLib;
   inherit (lib.lists) optionals;
   inherit (lib.strings) concatStringsSep;
 
@@ -22,15 +23,21 @@ let
     cuda_cccl
     cuda_cudart
     cuda_nvcc
+    cuda_nvml_dev
     cudaFlags
+    nccl
     ;
+
+  stdenv = throw "Use effectiveStdenv instead";
+  effectiveStdenv = if enableCuda then cudaPackages.backendStdenv else inputs.stdenv;
 in
-stdenv.mkDerivation (finalAttrs: {
+effectiveStdenv.mkDerivation (finalAttrs: {
   __structuredAttrs = true;
-  # TODO: When strictDeps is enabled, the CUDA build fails during configurePhase because it can't find all the CUDA
-  # dependencies. As such, we hold off on enabling strictDeps until CUDA compilation works.
-  # https://github.com/openucx/ucc/blob/0c0fc21559835044ab107199e334f7157d6a0d3d/config/m4/cuda.m4
-  strictDeps = false;
+  # TODO(@connorbaker):
+  # When strictDeps is enabled, `cuda_nvcc` is required as the argument to `--with-cuda` in `configureFlags` or else
+  # configurePhase fails with `checking for cuda_runtime.h... no`.
+  # This is odd, especially given `cuda_runtime.h` is provided by `cuda_cudart.dev`, which is already in `buildInputs`.
+  strictDeps = true;
 
   pname = "ucc";
   version = "1.3.0";
@@ -55,7 +62,7 @@ stdenv.mkDerivation (finalAttrs: {
       substituteInPlace "$comp" \
         --replace-quiet \
           "/bin/bash" \
-          "${stdenv.shell}"
+          "${effectiveStdenv.shell}"
     done
   '';
 
@@ -70,7 +77,19 @@ stdenv.mkDerivation (finalAttrs: {
     ++ optionals enableCuda [
       cuda_cccl
       cuda_cudart
+      cuda_nvml_dev
+      nccl
     ];
+
+  # NOTE: With `__structuredAttrs` enabled, `LDFLAGS` must be set under `env` so it is assured to be a string;
+  # otherwise, we might have forgotten to convert it to a string and Nix would make LDFLAGS a shell variable
+  # referring to an array!
+  env.LDFLAGS = builtins.toString (
+    optionals enableCuda [
+      # Fake libnvidia-ml.so (the real one is deployed impurely)
+      "-L${getLib cuda_nvml_dev}/lib/stubs"
+    ]
+  );
 
   preConfigure = ''
     ./autogen.sh
@@ -81,7 +100,7 @@ stdenv.mkDerivation (finalAttrs: {
     ++ optionals enableSse42 [ "--with-sse42" ]
     ++ optionals enableAvx [ "--with-avx" ]
     ++ optionals enableCuda [
-      "--with-cuda=${cuda_cudart}"
+      "--with-cuda=${cuda_nvcc}"
       "--with-nvcc-gencode=${concatStringsSep " " cudaFlags.gencode}"
     ];
 

--- a/pkgs/by-name/uc/ucc/package.nix
+++ b/pkgs/by-name/uc/ucc/package.nix
@@ -1,28 +1,45 @@
 {
-  stdenv,
-  lib,
-  fetchFromGitHub,
-  libtool,
-  automake,
   autoconf,
-  ucx,
+  automake,
   config,
-  enableCuda ? config.cudaSupport,
   cudaPackages,
+  fetchFromGitHub,
+  lib,
+  libtool,
+  stdenv,
+  ucx,
+  # Configuration options
   enableAvx ? stdenv.hostPlatform.avxSupport,
+  enableCuda ? config.cudaSupport,
   enableSse41 ? stdenv.hostPlatform.sse4_1Support,
   enableSse42 ? stdenv.hostPlatform.sse4_2Support,
 }:
+let
+  inherit (lib.lists) optionals;
+  inherit (lib.strings) concatStringsSep;
 
-stdenv.mkDerivation rec {
+  inherit (cudaPackages)
+    cuda_cccl
+    cuda_cudart
+    cuda_nvcc
+    cudaFlags
+    ;
+in
+stdenv.mkDerivation (finalAttrs: {
+  __structuredAttrs = true;
+  # TODO: When strictDeps is enabled, the CUDA build fails during configurePhase because it can't find all the CUDA
+  # dependencies. As such, we hold off on enabling strictDeps until CUDA compilation works.
+  # https://github.com/openucx/ucc/blob/0c0fc21559835044ab107199e334f7157d6a0d3d/config/m4/cuda.m4
+  strictDeps = false;
+
   pname = "ucc";
   version = "1.3.0";
 
   src = fetchFromGitHub {
     owner = "openucx";
     repo = "ucc";
-    rev = "v${version}";
-    sha256 = "sha256-xcJLYktkxNK2ewWRgm8zH/dMaIoI+9JexuswXi7MpAU=";
+    tag = "v${finalAttrs.version}";
+    hash = "sha256-xcJLYktkxNK2ewWRgm8zH/dMaIoI+9JexuswXi7MpAU=";
   };
 
   outputs = [
@@ -32,44 +49,46 @@ stdenv.mkDerivation rec {
 
   enableParallelBuilding = true;
 
+  # NOTE: We use --replace-quiet because not all Makefile.am files contain /bin/bash.
   postPatch = ''
-
     for comp in $(find src/components -name Makefile.am); do
-      substituteInPlace $comp \
-        --replace "/bin/bash" "${stdenv.shell}"
+      substituteInPlace "$comp" \
+        --replace-quiet \
+          "/bin/bash" \
+          "${stdenv.shell}"
     done
   '';
 
   nativeBuildInputs = [
-    libtool
-    automake
     autoconf
-  ] ++ lib.optionals enableCuda [ cudaPackages.cuda_nvcc ];
+    automake
+    libtool
+  ] ++ optionals enableCuda [ cuda_nvcc ];
+
   buildInputs =
     [ ucx ]
-    ++ lib.optionals enableCuda [
-      cudaPackages.cuda_cccl
-      cudaPackages.cuda_cudart
+    ++ optionals enableCuda [
+      cuda_cccl
+      cuda_cudart
     ];
 
-  preConfigure =
-    ''
-      ./autogen.sh
-    ''
-    + lib.optionalString enableCuda ''
-      configureFlagsArray+=( "--with-nvcc-gencode=${builtins.concatStringsSep " " cudaPackages.cudaFlags.gencode}" )
-    '';
+  preConfigure = ''
+    ./autogen.sh
+  '';
+
   configureFlags =
-    [ ]
-    ++ lib.optional enableSse41 "--with-sse41"
-    ++ lib.optional enableSse42 "--with-sse42"
-    ++ lib.optional enableAvx "--with-avx"
-    ++ lib.optional enableCuda "--with-cuda=${cudaPackages.cuda_cudart}";
+    optionals enableSse41 [ "--with-sse41" ]
+    ++ optionals enableSse42 [ "--with-sse42" ]
+    ++ optionals enableAvx [ "--with-avx" ]
+    ++ optionals enableCuda [
+      "--with-cuda=${cuda_cudart}"
+      "--with-nvcc-gencode=${concatStringsSep " " cudaFlags.gencode}"
+    ];
 
   postInstall = ''
-    find $out/lib/ -name "*.la" -exec rm -f \{} \;
+    find "$out/lib/" -name "*.la" -exec rm -f \{} \;
 
-    moveToOutput bin/ucc_info $dev
+    moveToOutput bin/ucc_info "$dev"
   '';
 
   meta = with lib; {
@@ -79,4 +98,4 @@ stdenv.mkDerivation rec {
     maintainers = [ maintainers.markuskowa ];
     platforms = platforms.linux;
   };
-}
+})

